### PR TITLE
[packaging] Bump golang-crossbuild version debian8

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

Bump version:
* from debian7 -> debian8.  <-- Need to clarify this

## Why is it important?

As discussed offline to use debian 8 images

## Failures

* x-pack/auditbeat packaging failed

Command = `PACKAGES='+all linux/amd64 linux/386 linux/arm64 linux/armv7 windows/amd64 windows/386' mage package`

Error
```
[2021-02-05T14:02:50.638Z] >> Building using: cmd='build/mage-linux-amd64 golangCrossBuild', env=[CC=gcc, CXX=g++, GOARCH=386, GOARM=, GOOS=linux, PLATFORM_ID=linux-386]
[2021-02-05T14:03:02.877Z] debconf: delaying package configuration, since apt-utils is not installed
[2021-02-05T14:03:17.797Z] W: GPG error: http://archive.debian.org jessie Release: The following signatures were invalid: KEYEXPIRED 1587841717
[2021-02-05T14:03:20.346Z] W: GPG error: http://archive.debian.org jessie Release: The following signatures were invalid: KEYEXPIRED 1587841717
[2021-02-05T14:03:30.356Z] E: There are problems and -y was used without --force-yes
[2021-02-05T14:03:30.356Z] Error: running "apt-get install -y --no-install-recommends librpm-dev" failed with exit code 100
[2021-02-05T14:03:30.356Z] Error: failed building for linux/amd64: exit status 100
[2021-02-05T14:03:30.356Z] failed building for linux/amd64: exit status 100
[2021-02-05T14:03:35.634Z] E: There are problems and -y was used without --force-yes
[2021-02-05T14:03:35.634Z] Error: running "apt-get install -y --no-install-recommends librpm-dev:i386" failed with exit code 100
[2021-02-05T14:03:35.634Z] Error: failed building for linux/386: exit status 100
[2021-02-05T14:03:35.634Z] failed building for linux/386: exit status 100
[2021-02-05T14:04:02.201Z] # crypto/x509
[2021-02-05T14:04:02.201Z] ld: warning: object file (/tmp/go-build127571197/b115/_cgo_main.o) was built for newer OSX version (10.11) than being linked (10.10)
[2021-02-05T14:04:02.201Z] ld: warning: object file (/tmp/go-build127571197/b115/_x001.o) was built for newer OSX version (10.11) than being linked (10.10)
[2021-02-05T14:04:02.201Z] ld: warning: object file (/tmp/go-build127571197/b115/_x002.o) was built for newer OSX version (10.11) than being linked (10.10)
[2021-02-05T14:07:23.765Z] # github.com/elastic/beats/v7/x-pack/auditbeat
[2021-02-05T14:07:23.765Z] ld: warning: object file (/tmp/go-link-221104519/000029.o) was built for newer OSX version (10.11) than being linked (10.10)
[2021-02-05T14:07:23.765Z] ld: warning: object file (/tmp/go-link-221104519/000030.o) was built for newer OSX version (10.11) than being linked (10.10)
[2021-02-05T14:07:23.765Z] package ran for 5m12.397231978s
[2021-02-05T14:07:23.765Z] Error: running "docker run --env EXEC_UID=1160 --env EXEC_GID=1161 -v /var/lib/jenkins/workspace/Beats_packaging_PR-23872/pkg/mod:/go/pkg/mod:ro --rm --env GOFLAGS=-mod=readonly --env MAGEFILE_VERBOSE= --env MAGEFILE_TIMEOUT= --env SNAPSHOT=true --env DEV=true -v /var/lib/jenkins/workspace/Beats_packaging_PR-23872/src/github.com/elastic/beats:/go/src/github.com/elastic/beats -w /go/src/github.com/elastic/beats/x-pack/auditbeat docker.elastic.co/beats-dev/golang-crossbuild:1.15.7-main-debian8 --build-cmd build/mage-linux-amd64 golangCrossBuild -p linux/amd64" failed with exit code 1
[2021-02-05T14:07:23.765Z] running "docker run --env EXEC_UID=1160 --env EXEC_GID=1161 -v /var/lib/jenkins/workspace/Beats_packaging_PR-23872/pkg/mod:/go/pkg/mod:ro --rm --env GOFLAGS=-mod=readonly --env MAGEFILE_VERBOSE= --env MAGEFILE_TIMEOUT= --env SNAPSHOT=true --env DEV=true -v /var/lib/jenkins/workspace/Beats_packaging_PR-23872/src/github.com/elastic/beats:/go/src/github.com/elastic/beats -w /go/src/github.com/elastic/beats/x-pack/auditbeat docker.elastic.co/beats-dev/golang-crossbuild:1.15.7-main-debian8 --build-cmd build/mage-linux-amd64 golangCrossBuild -p linux/386" failed with exit code 1
script returned exit code 1
```

### CODE REVIEW

**Review by mdelapenya — APPROVED** [review]

File: dev-tools/mage/crossbuild.go
Comment (kuisathaverat) [inline dev-tools/mage/crossbuild.go:197]: in order to keep compatibility with glibc 2.13 we should keep using debian7 to build i386/amd64 binaries

Comment (urso) [reply dev-tools/mage/crossbuild.go:197]: 2.19 might be fine. See glibc compliance tracker: https://abi-laboratory.pro/?view=timeline&l=glibc

If not we might see beats-tester complain if we test with debian 7.

**Review by cachedout — APPROVED** [review]

File: .ci/packer_cache.sh
Comment (v1v) [inline .ci/packer_cache.sh:67]: We keep the debian-7 since 7.x branches might use it

**Review by urso — APPROVED** [review]

**Comment by v1v** [comment]
@jsoriano , I might need some help with the existing failures when packaging the `x-pack/auditbeat` on Linux. I saw your changes in https://github.com/elastic/beats/pull/21239  a few months ago, any hints what's happening?

Logs -> https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-23872/runs/4/steps/12600/log/?start=0

**Comment by jsoriano** [comment]
> @jsoriano , I might need some help with the existing failures when packaging the `x-pack/auditbeat` on Linux. I saw your changes in #21239 a few months ago, any hints what's happening?
> 
> Logs -> https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-23872/runs/4/steps/12600/log/?start=0

@v1v it seems that the base image has an incorrect `sources.list`, or PGP keys for the repos listed there.

It seems that jessie is still available under http://http.debian.net/debian, so maybe we should use this repository instead of http://archive.debian.org/debian.

Official `debian:8` image uses http://deb.debian.org/debian.

**Comment by v1v** [comment]
@jsoriano , sorry that I missed this PR for a while.

> It seems that jessie is still available under http://http.debian.net/debian, so maybe we should use this repository instead of http://archive.debian.org/debian.

IIUC, Jessie Long Term Support (LTS) ended in June 2020. https://github.com/elastic/golang-crossbuild/pull/72 was the one that moved to debian-9 and changed to use `archive` for debian-8 and debian-7

I don't know the policy regarding what sources should be used, but I guess the changes to be done should be in the `golang-crossbuild`, right?

**Comment by jsoriano** [comment]
> I don't know the policy regarding what sources should be used, but I guess the changes to be done should be in the `golang-crossbuild`, right?

Yes, I think so. As jessie is not maintained anymore maybe an approach is to remove all unavailable sources (or all of them as they are not being updated anymore and no more updates will be received).

**Comment by v1v** [comment]
> Yes, I think so. As jessie is not maintained anymore maybe an approach is to remove all unavailable sources (or all of them as they are not being updated anymore and no more updates will be received).

I guess, that's something to be done in the `golang-crossbuild` project, so maybe that's something the Beats team need to address it?

For the time being, do you think we can jump from `debian-7` to `debian-9`? Since debian-8 is already EOL. The question is `debian 9 uses glibc 2.24` so I don't know if that could be an issue

**Comment by jsoriano** [comment]
@v1v was it only happening with auditbeat builds?

Since Ivan included the precompiled libpcap (https://github.com/elastic/beats/pull/25769 https://github.com/elastic/golang-crossbuild/pull/91) auditbeat build got simplified, could you update the branch to see if these problems are solved?

**Comment by v1v** [comment]
> Since Ivan included the precompiled libpcap (#25769 elastic/golang-crossbuild#91) auditbeat build got simplified, could you update the branch to see if these problems are solved?

I already triggered a build that should have those changes, let's see how it goes :) 

Thanks for the information :)

**Comment by v1v** [comment]
@jsoriano , same problem.

**Comment by kuisathaverat** [comment]
IIRC libpcap is used on Packetbeat, I am not sure Auditbeat uses it.

**Comment by kuisathaverat** [comment]
It is trying to install packages and failed, we have to get rid of packages installation during the build, if we need those packages we should install them in the image, indeed I think they are installed in other architectures https://github.com/elastic/golang-crossbuild/blob/master/go1.16/mips/Dockerfile.tmpl#L24-L36.

```
[2021-07-08T18:07:07.565Z] Error: running "apt-get install -y --no-install-recommends librpm-dev" failed with exit code 100
[2021-07-08T18:07:07.565Z] Error: failed building for linux/amd64: exit status 100
[2021-07-08T18:07:07.565Z] failed building for linux/amd64: exit status 100
[2021-07-08T18:07:07.565Z] 43897a096976: Pull complete
[2021-07-08T18:07:07.565Z] E: There are problems and -y was used without --force-yes
[2021-07-08T18:07:07.565Z] Error: running "apt-get install -y --no-install-recommends librpm-dev:i386" failed with exit code 100
[2021-07-08T18:07:07.565Z] Error: failed building for linux/386: exit status 100
[2021-07-08T18:07:07.565Z] failed building for linux/386: exit status 100
```

File: dev-tools/mage/crossbuild.go
Comment (v1v) [reply dev-tools/mage/crossbuild.go:225]:
```suggestion
		tagSuffix = "main-debian9"
```

**Comment by kuisathaverat** [comment]
I have added the libraries to the goland-crossbuild Docker images at https://github.com/elastic/golang-crossbuild/pull/117

**Comment by kuisathaverat** [comment]
I just merged elastic/golang-crossbuild#117, when [the master build ends](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fgolang-crossbuild-mbp/detail/master/35/pipeline) the new Docker images would be available with the packages installed. I do not remember if there's something to remove from the package.go in Auditbeat but I guess so to avoid future issues.

File: dev-tools/mage/crossbuild.go
Comment (kuisathaverat) [inline dev-tools/mage/crossbuild.go:225]:
```suggestion
		tagSuffix = "main-debian8"
```

**Review by kuisathaverat — APPROVED** [review]

**Comment by kuisathaverat** [comment]
we should get rid of all this code https://github.com/elastic/beats/blob/23e4403ae093fcc8f7905345cad2c7ad256976d8/x-pack/auditbeat/magefile.go#L128-L218

**Comment by kuisathaverat** [comment]
it will fail :D you have to remove the call to the function too https://github.com/elastic/beats/blob/23e4403ae093fcc8f7905345cad2c7ad256976d8/x-pack/auditbeat/magefile.go#L47-L49

**Comment by v1v** [comment]
Test failures related to some regression, fix is coming with -> https://github.com/elastic/beats/pull/26972

So let's wait until it gets merged

**Comment by v1v** [comment]
>  I guess that old images continue to be available to keep compatibility with old OSs during 7.x? 

Yes, `golang-crossbuild` images are never removed as far as I know.

I'll keep this PR open for a couple of more days, @ruflin and @andrewkroh please feel free to comment if needed otherwise I'll assume we are good to merge. Thanks

**Comment by v1v** [comment]
I'll fix the conflict and then if I don't hear from @andrewkroh or any others by Thursday 2nd Dec ( in the afternoon CEST ) then I'll proceed with the merge

**Comment by andrewkroh** [comment]
> could this have any effect on the compatibility of some of the beats on older platforms? If yes, do we still support these?

CentOS 6,  which has glibc 2.12, is officially supported in 7.16. beats-tester no longer executes on RHEL/CentOS 6 (as of https://github.com/elastic/beats-tester/pull/193). So unless manual testing is performed I doubt we will catch issues.

While integrating journald support into Filebeat I had to be careful not to accidentally bump the glibc requirement of Filebeat. To ensure that the glibc requirements remained at <= 2.12 I created a test that analyzed the ELF binary produced during the golangCrossBuild step. I only added this test to Filebeat because this was the only target where I was changing the golang-crossbuild base image.

But now that we're changing the base image for all Beats and we don't have runtime testing on CentOS 6, I think it would be worth the effort to add this static test into the build of Beats that are required to support CentOS 6.

https://github.com/elastic/beats/blob/6b417428d30971c75469651419cf39874189f268/dev-tools/mage/pkg.go#L247-L257

**Comment by v1v** [comment]
Thanks @andrewkroh.

IIUC, we don't test on CentOS6 therefore we could still move to debian8? If so, who can help on adding the TestLinuxForCentosGLIBC static test into the build of Beats?

Thanks

**Comment by andrewkroh** [comment]
Yes, I think we are safe to move to the debian8 based build images provided that we add the TestLinuxForCentosGLIBC static test to check the binaries coming out of the build.

That test will look for any glibc symbols referenced in the binary and check their version. For example if a 2.13 glibc symbol is referenced then the check will fail. And if I wrote the test correctly, that same binary would fail if you tried to actually use it on CentOS 6 since it only has glibc 2.12.

**Comment by andrewkroh** [comment]
> Yes, ... provided that we add the TestLinuxForCentosGLIBC static test to check the binaries coming out of the build.

This PR proceeded without adding the tests and unfortunately Metricbeat and Packetbeat no longer run on CentOS 6 for 7.17 and 8.0. If you build the Dockerfile below you see the problem that occurs at runtime.

> usr/share/metricbeat/bin/metricbeat: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by /usr/share/metricbeat/bin/metricbeat)
>
> /usr/share/packetbeat/bin/packetbeat: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by /usr/share/packetbeat/bin/packetbeat)

```
# Dockerfile
FROM centos:6

ENV SNAPSHOT https://snapshots.elastic.co/7.17.0-fd761817/downloads/beats

RUN curl -O $SNAPSHOT/metricbeat/metricbeat-7.17.0-SNAPSHOT-x86_64.rpm \
      && rpm -ivh metricbeat-7.17.0-SNAPSHOT-x86_64.rpm

RUN curl -O $SNAPSHOT/packetbeat/packetbeat-7.17.0-SNAPSHOT-x86_64.rpm \
      && rpm -ivh packetbeat-7.17.0-SNAPSHOT-x86_64.rpm

RUN /usr/share/metricbeat/bin/metricbeat version \
    || /usr/share/packetbeat/bin/packetbeat version
```

**Comment by v1v** [comment]
:/

I actually misunderstood that we were safe to move, what do we do here?

**Comment by andrewkroh** [comment]
@v1v is reverting the changes so we can get those Beats back to a working state for CentOS 6. Currently the 7.17.0 release is affected and the 8.0.0 RCs.

Once we get back to a working state we can discuss how to safely move forward with this change to transition off of debian7 for building.

**Comment by andrewkroh** [comment]
I can't find a decision regarding 8.0 support for CentOS 6 in Beats/Agent. But if for 8.0 we are dropping support for Beats and Agent on CentOS/RHEL 6 then the changes might be OK for 8.0 and master.

We should document the minimum supported GLIBC version for Beats and add the tests I mentioned for it. That will prevent us from accidentally moving to a new builder image in the future that goes beyond our minimum glibc version.

That minimum supported GLIBC would basically be the oldest glibc version found in all the supported OSes.

**Comment by nimarezainia** [comment]
@andrewkroh CentOS EOL November 2020. with all honesty it should have been removed from our support matrix 6-12 months after that. No hard and fast rule about this but I'm trying to clean up our support matrix and set up a policy as such.

I deprecated this support from the matrix in 8.0. the 7.17 matrix was just published. I can try and see if that can be modified.

**Comment by jlind23** [comment]
@nimarezainia does it mean that we are not supporting RHEL6 from 7.17 on?

**Comment by jlind23** [comment]
@v1v was it merged back to 7.17 and will it be part of a 7.17.1 release?

**Comment by nimarezainia** [comment]
> @nimarezainia does it mean that we are not supporting RHEL6 from 7.17 on?

yes correct. 7.17 is the last release.

**Comment by jlind23** [comment]
@v1v sorry I was talking about the revert.
Was it revert on 7.17 branch? If yes, then I'm good.

**Comment by nimarezainia** [comment]
thank you all. So we will let customers know that 7.17.1 is fixed.

**Pull request merged**
